### DISCUSSION TITLE
Problem on sealing procedure in webpack

### DISCUSSION THREAD

**Discussion by Anemone95** [webpack/webpack#18796] [Q&A] [answered]
When webpack generates a bundle, it puts some modules into a map object on the top, but it also puts some modules' functions and codes directly into the body part. Sometimes, these functions are not only application code but also 3rd party libraries:

For example, I have code like:
```
import React, { StrictMode } from "react";
import { createRoot } from "react-dom/client";
import "./styles.css";
import { BrowserRouter, Routes, Route } from "react-router-dom";

import {Greeting} from "./Home";


export default function App() {
  return (
    <BrowserRouter>
      <Routes>
        <Route path="/" element={<Greeting />} />
      </Routes>
    </BrowserRouter>
  );
}

const root = createRoot(document.getElementById("root"));
root.render(
  <StrictMode>
    <App />
  </StrictMode>
);

```

The compiled code is like:
```
(() => {
    "use strict";
    var e = {
            730: (e, t, n) => {
                var r = n(43),
                ...
           }
     }
    /*...*/

    function ce(e) {
        let {
            basename: t,
            children: n,
            future: r,
            window: a
        } = e, o = l.useRef();
        var u;
        null == o.current && (o.current = (void 0 === (u = {
            window: a,
            v5Compat: !0
        }) && (u = {}), g((function(e, t) {
            let {
                pathname: n,
                search: r,
                hash: l
            } = e.location;
            return h("", {
                pathname: n,
                search: r,
                hash: l
            }, t.state && t.state.usr || null, t.state && t.state.key || "default")
        }), (function(e, t) {
            return "string" === typeof t ? t : m(t)
        }), null, u)));
        let i = o.current,
            [s, c] = l.useState({
                action: i.action,
                location: i.location
            }),
            {
                v7_startTransition: f
            } = r || {},
            d = l.useCallback((e => {
                f && se ? se((() => c(e))) : c(e)
            }), [c, f]);
        return l.useLayoutEffect((() => i.listen(d)), [i, d]), l.createElement(oe, {
            basename: t,
            children: n,
            location: s.location,
            navigationType: s.action,
            navigator: i,
            future: r
        })
    }
    "undefined" !== typeof window && "undefined" !== typeof window.document && window.document.createElement;
    var fe, de;
    (function(e) {
        e.UseScrollRestoration = "useScrollRestoration", e.UseSubmit = "useSubmit", e.UseSubmitFetcher = "useSubmitFetcher", e.UseFetcher = "useFetcher", e.useViewTransitionState = "useViewTransitionState"
    })(fe || (fe = {})),
    function(e) {
        e.UseFetcher = "useFetcher", e.UseFetchers = "useFetchers", e.UseScrollRestoration = "useScrollRestoration"
    }(de || (de = {}));
    var pe = n(579);
    /*...*/

    function me() {
        return (0, pe.jsx)(ce, {
            children: (0, pe.jsx)(ue, {
                children: (0, pe.jsx)(ae, {
                    path: "/",
                    element: (0, pe.jsx)(he, {})
                })
            })
        })
    }(0, o.H)(document.getElementById("root")).render((0, pe.jsx)(l.StrictMode, {
        children: (0, pe.jsx)(me, {})
    }))
```
The function `ce` does not belong to my code but `react-router-dom`. So the problem is, to my expectation, `react-router-dom` code should be put into map `e` and use the requiring mechanism to get. But it stays in the main part, the same place as the application code stays.

So why this could happen and which kind of library or function can be put in the "main" part? And which part of the code of webpack can I look at if I want to learn the details of this procedure?

**Comment by alexander-akait** [comment]
Because some modules can be inlined, some of them not, so when it is possible, webpack doesn't it

**Reply by Anemone95** [reply]
I guess so. "some modules can be inlined, some of them not", so do you know where the code related to this logic is?

**Reply by alexander-akait** [reply]
It is not easy - https://github.com/webpack/webpack/blob/main/lib/optimize/ConcatenatedModule.js

**Reply by Anemone95** [reply]
I see. And I guess there isn't any document describing how this logic is? So the only way to know the details is to read this code?

**Reply by alexander-akait** [reply]
Yeah, logic is complex, better to debug and read this code

**Reply by Anemone95** [reply]
Got it. Thanks!

**Comment by absidue** [comment, marked as answer]
@Anemone95 The general rule is ES Modules go in the main part, because their exports are static, which means they can't change at runtime, so it is safe for webpack to inline them/merge them together with other modules. CommonJS goes into the map at the top, because CommonJS allows dynamic exports, that means you could change them at runtime, so webpack has to simulate the imports at runtime (you'll notice that all accesses to that map are function calls, not direct access), which is why they go in that map.

**Reply by Anemone95** [reply]
Yes. After one day of debugging, I think you are right!

**Reply by Anemone95** [reply]
But what will happen when the CommonJS module requires an ESModule? I guess this time ESModule can't be hoisted in main part?

**Reply by absidue** [reply]
That is one of the exceptions yes. As previously stated the logic is quite complex.
